# Pertemuan 2 — Struktur Data Python, NumPy & Pandas

**Mata Kuliah:** Pengantar Data Science (Kode: 200302305)  
**Program Studi:** PJJ Informatika  
**Semester:** 4  
**Dosen:** Syahid Abdullah, S.Si, M.Kom

---

**Nama :** Junior Dany Wibisono  
**NIM  :** 250401020098  
**Angkatan:** 2025  
**Tanggal:** 26 April 2026

---

## Tujuan Notebook
Notebook ini dibuat sebagai aktivitas hands-on Pertemuan 2 untuk:
1. Memahami dan membedakan struktur data Python: list, tuple, dictionary, dan set
2. Membuat dan memanipulasi NumPy ndarray serta menerapkan operasi vektorisasi
3. Menggunakan Pandas Series dan DataFrame untuk analisis data
4. Mengeksplorasi dataset Titanic dan menjawab pertanyaan analitik Q1–Q6

---
## ⚙️ Setup — Import Library

In [11]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

print('=' * 45)
print('  ✅ Library berhasil diimport!')
print('=' * 45)
print(f'  Pandas  : v{pd.__version__}')
print(f'  NumPy   : v{np.__version__}')

  ✅ Library berhasil diimport!
  Pandas  : v2.2.2
  NumPy   : v2.0.2


---
## 📥 Load Dataset Titanic

In [12]:
# Memuat dataset Titanic dari GitHub (raw CSV)
URL = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

df = pd.read_csv(URL)

print('✅ Dataset Titanic berhasil dimuat!')
print(f'   Shape : {df.shape}')
print()
print('5 Baris Pertama Dataset:')
df.head()

✅ Dataset Titanic berhasil dimuat!
   Shape : (891, 12)

5 Baris Pertama Dataset:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


---
## 🔍 Q1 — Jumlah Baris, Kolom, dan Daftar Kolom Dataset

In [13]:
print('=' * 55)
print('  Q1: DIMENSI DAN INFORMASI KOLOM DATASET TITANIC')
print('=' * 55)

n_baris, n_kolom = df.shape

print(f'\n📐 Dimensi Dataset   : {df.shape}')
print(f'   ├── Jumlah Baris  : {n_baris} baris (record penumpang)')
print(f'   └── Jumlah Kolom  : {n_kolom} kolom (fitur)')

print(f'\n📋 Daftar Kolom Dataset:')
print(f'   {"No.":<4} {"Nama Kolom":<16} {"Tipe Data":<12} Keterangan')
print(f'   {"---":<4} {"-----------":<16} {"---------":<12} ----------')

keterangan = {
    'PassengerId' : 'ID unik penumpang',
    'Survived'    : 'Status selamat (0=Tidak, 1=Ya) ← Target',
    'Pclass'      : 'Kelas tiket (1=1st, 2=2nd, 3=3rd)',
    'Name'        : 'Nama lengkap penumpang',
    'Sex'         : 'Jenis kelamin (male/female)',
    'Age'         : 'Usia penumpang (tahun)',
    'SibSp'       : 'Jumlah saudara/pasangan di kapal',
    'Parch'       : 'Jumlah orang tua/anak di kapal',
    'Ticket'      : 'Nomor tiket',
    'Fare'        : 'Harga tiket (£)',
    'Cabin'       : 'Nomor kabin',
    'Embarked'    : 'Pelabuhan embarkasi (C=Cherbourg, Q=Queenstown, S=Southampton)'
}

for i, (col, dtype) in enumerate(zip(df.columns, df.dtypes), 1):
    print(f'   {i:<4} {col:<16} {str(dtype):<12} {keterangan.get(col, "-")}')

print()
print('📌 Info Detail (df.info()):')
df.info()

  Q1: DIMENSI DAN INFORMASI KOLOM DATASET TITANIC

📐 Dimensi Dataset   : (891, 12)
   ├── Jumlah Baris  : 891 baris (record penumpang)
   └── Jumlah Kolom  : 12 kolom (fitur)

📋 Daftar Kolom Dataset:
   No.  Nama Kolom       Tipe Data    Keterangan
   ---  -----------      ---------    ----------
   1    PassengerId      int64        ID unik penumpang
   2    Survived         int64        Status selamat (0=Tidak, 1=Ya) ← Target
   3    Pclass           int64        Kelas tiket (1=1st, 2=2nd, 3=3rd)
   4    Name             object       Nama lengkap penumpang
   5    Sex              object       Jenis kelamin (male/female)
   6    Age              float64      Usia penumpang (tahun)
   7    SibSp            int64        Jumlah saudara/pasangan di kapal
   8    Parch            int64        Jumlah orang tua/anak di kapal
   9    Ticket           object       Nomor tiket
   10   Fare             float64      Harga tiket (£)
   11   Cabin            object       Nomor kabin
   12   Embark

### 📝 Interpretasi Q1

Dataset Titanic memiliki **891 baris** (mewakili 891 penumpang) dan **12 kolom** (fitur). Kolom-kolom tersebut mencakup:
- **Identitas:** `PassengerId`, `Name`, `Ticket`
- **Demografi:** `Sex`, `Age`, `SibSp` (saudara/pasangan), `Parch` (orang tua/anak)
- **Perjalanan:** `Pclass`, `Fare`, `Cabin`, `Embarked`
- **Label/Target:** `Survived` (0 = tidak selamat, 1 = selamat)

---
## 🔍 Q2 — Kolom dengan Missing Values

In [14]:
print('=' * 55)
print('  Q2: ANALISIS MISSING VALUES PER KOLOM')
print('=' * 55)

# Hitung jumlah dan proporsi missing values
missing_count = df.isnull().sum()
missing_pct   = df.isnull().mean() * 100

# Buat DataFrame ringkasan
missing_df = pd.DataFrame({
    'Jumlah_Missing' : missing_count,
    'Proporsi_%'     : missing_pct.round(2)
}).sort_values('Jumlah_Missing', ascending=False)

# Kolom dengan missing values
ada_missing = missing_df[missing_df['Jumlah_Missing'] > 0].copy()
ada_missing['Status'] = ada_missing['Proporsi_%'].apply(
    lambda x: '🔴 KRITIS (>50%)' if x >= 50 else ('🟡 SEDANG (10-50%)' if x >= 10 else '🟢 MINOR (<10%)')
)

print('\n📊 Kolom DENGAN missing values:')
print(ada_missing.to_string())

# Kolom tanpa missing
tidak_missing = missing_df[missing_df['Jumlah_Missing'] == 0].index.tolist()
print(f'\n✅ Kolom TANPA missing values ({len(tidak_missing)} kolom):')
print(f'   {", ".join(tidak_missing)}')

# Total
total_missing = df.isnull().sum().sum()
total_cells   = df.shape[0] * df.shape[1]
print(f'\n📈 Total sel kosong  : {total_missing} dari {total_cells:,} sel ({total_missing/total_cells*100:.2f}%)')

print('\n📋 Rekomendasi Penanganan:')
print('   • Cabin    → Drop kolom (>77% kosong, tidak informatif)')
print('   • Age      → Imputasi dengan nilai median (±19.87% kosong)')
print('   • Embarked → Imputasi dengan modus / nilai terbanyak (0.22%)')

  Q2: ANALISIS MISSING VALUES PER KOLOM

📊 Kolom DENGAN missing values:
          Jumlah_Missing  Proporsi_%             Status
Cabin                687       77.10    🔴 KRITIS (>50%)
Age                  177       19.87  🟡 SEDANG (10-50%)
Embarked               2        0.22     🟢 MINOR (<10%)

✅ Kolom TANPA missing values (9 kolom):
   PassengerId, Name, Pclass, Survived, Sex, Parch, SibSp, Fare, Ticket

📈 Total sel kosong  : 866 dari 10,692 sel (8.10%)

📋 Rekomendasi Penanganan:
   • Cabin    → Drop kolom (>77% kosong, tidak informatif)
   • Age      → Imputasi dengan nilai median (±19.87% kosong)
   • Embarked → Imputasi dengan modus / nilai terbanyak (0.22%)


### 📝 Interpretasi Q2

Terdapat **3 kolom** yang memiliki missing values:

| Kolom | Missing | Proporsi | Rekomendasi |
|-------|---------|----------|-------------|
| `Cabin` | 687 | 77.10% | 🔴 Drop — terlalu banyak kosong |
| `Age` | 177 | 19.87% | 🟡 Imputasi dengan median |
| `Embarked` | 2 | 0.22% | 🟢 Imputasi dengan modus ('S') |

Kolom **`Cabin`** memiliki missing value paling parah (77.10%), sehingga biasanya didrop dalam analisis lanjutan. Penanganan missing values adalah langkah penting sebelum pemodelan Machine Learning.

---
## 🔍 Q3 — Persentase Penumpang yang Selamat

In [15]:
print('=' * 55)
print('  Q3: PERSENTASE PENUMPANG YANG SELAMAT')
print('=' * 55)

# ─── Menggunakan NumPy ──────────────────────────────────
survived_arr  = df['Survived'].to_numpy()          # konversi ke NumPy array
n_total       = len(survived_arr)
n_selamat     = int(np.sum(survived_arr))          # sum nilai 1
n_meninggal   = n_total - n_selamat
pct_selamat   = np.mean(survived_arr) * 100        # mean = proporsi
pct_meninggal = 100 - pct_selamat

print(f'\n📊 Hasil Perhitungan (menggunakan NumPy):')
print(f'   Total penumpang       : {n_total:>5}')
print(f'   ✅ Selamat  (1)       : {n_selamat:>5}  →  {pct_selamat:.2f}%')
print(f'   ❌ Meninggal (0)      : {n_meninggal:>5}  →  {pct_meninggal:.2f}%')

# ─── Verifikasi dengan Pandas ───────────────────────────
print(f'\n📋 Verifikasi dengan Pandas (value_counts):')
vc     = df['Survived'].value_counts()
vc_pct = df['Survived'].value_counts(normalize=True) * 100
print(f'   Survived = 0 (Meninggal) : {vc[0]} penumpang  ({vc_pct[0]:.2f}%)')
print(f'   Survived = 1 (Selamat)   : {vc[1]} penumpang  ({vc_pct[1]:.2f}%)')

# ─── Statistik NumPy Array ──────────────────────────────
print(f'\n📐 Statistik Array Survived (NumPy):')
print(f'   np.sum  (total selamat) = {int(np.sum(survived_arr))}')
print(f'   np.mean (proporsi)      = {np.mean(survived_arr):.4f}')
print(f'   np.min                  = {int(np.min(survived_arr))}')
print(f'   np.max                  = {int(np.max(survived_arr))}')

  Q3: PERSENTASE PENUMPANG YANG SELAMAT

📊 Hasil Perhitungan (menggunakan NumPy):
   Total penumpang       :   891
   ✅ Selamat  (1)       :   342  →  38.38%
   ❌ Meninggal (0)      :   549  →  61.62%

📋 Verifikasi dengan Pandas (value_counts):
   Survived = 0 (Meninggal) : 549 penumpang  (61.62%)
   Survived = 1 (Selamat)   : 342 penumpang  (38.38%)

📐 Statistik Array Survived (NumPy):
   np.sum  (total selamat) = 342
   np.mean (proporsi)      = 0.3838
   np.min                  = 0
   np.max                  = 1


### 📝 Interpretasi Q3

Dari **891 penumpang** dalam dataset:
- **342 orang (38.38%)** berhasil selamat.
- **549 orang (61.62%)** tidak selamat.

Lebih dari separuh penumpang (61.62%) tidak berhasil diselamatkan — menggambarkan betapa dahsyatnya tragedi tenggelamnya kapal Titanic pada April 1912. Angka ini menjadi baseline penting untuk mengevaluasi model prediksi Machine Learning nantinya.

---
## 🔍 Q4 — Penumpang Wanita Kelas 1: Jumlah & Rata-rata Usia

In [16]:
print('=' * 55)
print('  Q4: ANALISIS PENUMPANG WANITA KELAS 1')
print('=' * 55)

# ─── Boolean Indexing ───────────────────────────────────
mask_wanita = df['Sex'] == 'female'
mask_kelas1 = df['Pclass'] == 1

wk1 = df[mask_wanita & mask_kelas1]   # filter wanita kelas 1

jumlah      = len(wk1)
selamat     = int(wk1['Survived'].sum())
pct_selamat = wk1['Survived'].mean() * 100
rata_usia   = wk1['Age'].mean()
min_usia    = wk1['Age'].min()
max_usia    = wk1['Age'].max()
med_usia    = wk1['Age'].median()

print(f"\n🎫 Filter Applied: Sex == 'female' AND Pclass == 1")
print()
print(f'   Jumlah penumpang wanita kelas 1 : {jumlah} orang')
print(f'   Jumlah yang selamat             : {selamat} orang')
print(f'   Tingkat keselamatan             : {pct_selamat:.2f}%')
print()
print(f'   Statistik Usia:')
print(f'   ├── Rata-rata (mean)  : {rata_usia:.2f} tahun')
print(f'   ├── Median            : {med_usia:.2f} tahun')
print(f'   ├── Termuda (min)     : {min_usia:.2f} tahun')
print(f'   └── Tertua (max)      : {max_usia:.2f} tahun')

print('\n📋 Contoh data wanita kelas 1 (5 baris pertama):')
display(wk1[['Name','Age','Survived','Fare','Embarked']].head())

# Perbandingan konteks
print('\n📊 Perbandingan Tingkat Keselamatan (Sex × Pclass) dalam %:')
tabel = df.groupby(['Pclass','Sex'])['Survived'].mean().mul(100).round(2).unstack()
print(tabel.to_string())

  Q4: ANALISIS PENUMPANG WANITA KELAS 1

🎫 Filter Applied: Sex == 'female' AND Pclass == 1

   Jumlah penumpang wanita kelas 1 : 94 orang
   Jumlah yang selamat             : 91 orang
   Tingkat keselamatan             : 96.81%

   Statistik Usia:
   ├── Rata-rata (mean)  : 34.61 tahun
   ├── Median            : 35.00 tahun
   ├── Termuda (min)     : 2.00 tahun
   └── Tertua (max)      : 63.00 tahun

📋 Contoh data wanita kelas 1 (5 baris pertama):


,Name,Age,Survived,Fare,Embarked
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1,71.2833,C
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,53.1000,S
11,"Bonnell, Miss. Elizabeth",58.0,1,26.5500,S
31,"Spencer, Mrs. William Augustus (Marie Eugenie)",NaN,1,146.5208,C
52,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",49.0,1,76.7292,C



📊 Perbandingan Tingkat Keselamatan (Sex × Pclass) dalam %:
Sex     female   male
Pclass               
1        96.81  36.89
2        92.11  15.74
3        50.00  13.54


### 📝 Interpretasi Q4

Terdapat **94 penumpang wanita kelas 1** dalam dataset ini.
- **91 dari 94 orang (96.81%)** berhasil selamat — angka yang sangat tinggi.
- **Rata-rata usia** mereka adalah sekitar **34.61 tahun**.

Tingkat keselamatan hampir 97% ini mencerminkan dua faktor utama:
1. **Kebijakan evakuasi:** "Women and children first" diprioritaskan.
2. **Lokasi kabin:** Penumpang kelas 1 berada di dek atas, paling dekat dengan sekoci penyelamat.

---
## 🔍 Q5 — Tingkat Keselamatan per Kelas Penumpang (Groupby)

In [17]:
print('=' * 55)
print('  Q5: TINGKAT KESELAMATAN PER KELAS PENUMPANG')
print('=' * 55)

# ─── Groupby dasar ──────────────────────────────────────
print('\n📊 [1] Groupby dasar: mean Survived per Pclass')
basic = df.groupby('Pclass')['Survived'].mean().mul(100).round(2)
print(basic.to_frame('% Selamat').to_string())

# ─── Groupby dengan agg() lengkap ───────────────────────
print('\n📊 [2] Groupby dengan .agg() — multi-fungsi:')
hasil = df.groupby('Pclass').agg(
    Total_Penumpang = ('PassengerId', 'count'),
    Jumlah_Selamat  = ('Survived'   , 'sum'),
    Pct_Selamat     = ('Survived'   , 'mean'),
    Rata_Usia       = ('Age'        , 'mean'),
    Rata_Fare       = ('Fare'       , 'mean')
).round(3)
hasil['Pct_Selamat']     = (hasil['Pct_Selamat'] * 100).round(2)
hasil['Jumlah_Selamat']  = hasil['Jumlah_Selamat'].astype(int)
print(hasil.to_string())

# ─── Kesimpulan ─────────────────────────────────────────
kelas_terbaik  = hasil['Pct_Selamat'].idxmax()
kelas_terburuk = hasil['Pct_Selamat'].idxmin()
print(f'\n🏆 Kelas TERTINGGI : Kelas {kelas_terbaik} → {hasil.loc[kelas_terbaik, "Pct_Selamat"]}% selamat')
print(f'📉 Kelas TERENDAH  : Kelas {kelas_terburuk} → {hasil.loc[kelas_terburuk, "Pct_Selamat"]}% selamat')

# ─── Pivot Table ────────────────────────────────────────
print('\n📊 [3] Pivot Table — % Survived (Pclass × Sex):')
pivot = pd.pivot_table(
    df,
    values  = 'Survived',
    index   = 'Pclass',
    columns = 'Sex',
    aggfunc = 'mean'
).mul(100).round(2)
print(pivot.to_string())

  Q5: TINGKAT KESELAMATAN PER KELAS PENUMPANG

📊 [1] Groupby dasar: mean Survived per Pclass
        % Selamat
Pclass           
1           62.96
2           47.28
3           24.24

📊 [2] Groupby dengan .agg() — multi-fungsi:
        Total_Penumpang  Jumlah_Selamat  Pct_Selamat  Rata_Usia  Rata_Fare
Pclass                                                                    
1                   216             136         63.0     38.233     84.155
2                   184              87         47.3     29.878     20.662
3                   491             119         24.2     25.141     13.676

🏆 Kelas TERTINGGI : Kelas 1 → 63.0% selamat
📉 Kelas TERENDAH  : Kelas 3 → 24.2% selamat

📊 [3] Pivot Table — % Survived (Pclass × Sex):
Sex     female   male
Pclass               
1        96.81  36.89
2        92.11  15.74
3        50.00  13.54


### 📝 Interpretasi Q5

Berdasarkan `groupby('Pclass')['Survived'].mean()`:

| Kelas | Total | Selamat | % Selamat |
|-------|-------|---------|----------|
| **1** | 216 | 136 | **62.96%** 🏆 |
| **2** | 184 | 87 | **47.28%** |
| **3** | 491 | 119 | **24.24%** 📉 |

**Kelas 1** memiliki tingkat keselamatan **tertinggi (62.96%)**, hampir 3× lebih tinggi dari kelas 3 (24.24%). Hal ini sangat berkorelasi dengan lokasi kabin (dek atas = akses sekoci lebih mudah) dan status sosial-ekonomi penumpang.

---
## 🔍 Q6 — Analisis Tambahan

In [18]:
print('=' * 55)
print('  Q6A: TINGKAT KESELAMATAN BERDASARKAN KELOMPOK USIA')
print('=' * 55)

# Buat kolom kelompok usia
df_age = df.dropna(subset=['Age']).copy()
bins   = [0, 12, 17, 59, 100]
labels = ['Anak (<12)', 'Remaja (12-17)', 'Dewasa (18-59)', 'Senior (60+)']
df_age['KelompokUsia'] = pd.cut(df_age['Age'], bins=bins, labels=labels)

# Groupby kelompok usia
age_stats = df_age.groupby('KelompokUsia', observed=True).agg(
    Total    = ('Survived', 'count'),
    Selamat  = ('Survived', 'sum'),
    Pct      = ('Survived', 'mean')
)
age_stats['Pct']     = age_stats['Pct'].mul(100).round(2)
age_stats['Selamat'] = age_stats['Selamat'].astype(int)
age_stats.columns   = ['Total', 'Selamat', '% Selamat']
print()
print(age_stats.to_string())

  Q6A: TINGKAT KESELAMATAN BERDASARKAN KELOMPOK USIA

                Total  Selamat  % Selamat
KelompokUsia                             
Anak (<12)         69       40      57.97
Remaja (12-17)     44       21      47.73
Dewasa (18-59)    575      222      38.61
Senior (60+)       26        7      26.92


In [19]:
print('=' * 55)
print('  Q6B: TINGKAT KESELAMATAN BERDASARKAN UKURAN KELUARGA')
print('=' * 55)

# Buat kolom ukuran keluarga
df['UkKeluarga']  = df['SibSp'] + df['Parch']
bins2   = [-1, 0, 3, 6, 20]
labels2 = ['Sendiri (0)', 'Kecil (1-3)', 'Sedang (4-6)', 'Besar (7+)']
df['GrupKeluarga'] = pd.cut(df['UkKeluarga'], bins=bins2, labels=labels2)

# Groupby ukuran keluarga
fam_stats = df.groupby('GrupKeluarga', observed=True).agg(
    Total   = ('Survived', 'count'),
    Selamat = ('Survived', 'sum'),
    Pct     = ('Survived', 'mean')
)
fam_stats['Pct']     = fam_stats['Pct'].mul(100).round(2)
fam_stats['Selamat'] = fam_stats['Selamat'].astype(int)
fam_stats.columns   = ['Total', 'Selamat', '% Selamat']

print()
print(fam_stats.to_string())
print('\n📌 Penumpang dengan keluarga kecil (1-3 anggota) memiliki')
print('   peluang selamat lebih tinggi dibanding yang pergi sendiri.')

  Q6B: TINGKAT KESELAMATAN BERDASARKAN UKURAN KELUARGA

              Total  Selamat  % Selamat
GrupKeluarga                           
Sendiri (0)     537      163      30.35
Kecil (1-3)     292      169      57.88
Sedang (4-6)     49       10      20.41
Besar (7+)       13        0       0.00

📌 Penumpang dengan keluarga kecil (1-3 anggota) memiliki
   peluang selamat lebih tinggi dibanding yang pergi sendiri.


In [20]:
print('=' * 55)
print('  Q6C: KORELASI TARIF TIKET (FARE) & KESELAMATAN')
print('=' * 55)

# Rata-rata Fare per status selamat
fare_stats = df.groupby('Survived')['Fare'].agg(['mean','median','min','max']).round(2)
fare_stats.index = ['❌ Meninggal (0)', '✅ Selamat (1)']

print()
print(fare_stats.to_string())

# Korelasi numerik
korelasi_fare = df[['Fare','Survived']].corr().iloc[0,1]
korelasi_age  = df[['Age','Survived']].dropna().corr().iloc[0,1]

print(f'\n📊 Korelasi Pearson:')
print(f'   Fare ↔ Survived : {korelasi_fare:.4f}  (positif lemah)')
print(f'   Age  ↔ Survived : {korelasi_age:.4f}  (negatif sangat lemah)')

print(f'\n💡 Kesimpulan:')
print(f'   Semakin tinggi tarif tiket, sedikit lebih besar kemungkinan selamat.')
print(f'   Fare sangat berkaitan dengan Pclass sebagai variabel utama.')

  Q6C: KORELASI TARIF TIKET (FARE) & KESELAMATAN

                  mean  median  min     max
❌ Meninggal (0)  22.12    10.5  0.0  263.00
✅ Selamat (1)    48.40    26.0  0.0  512.33

📊 Korelasi Pearson:
   Fare ↔ Survived : 0.2573  (positif lemah)
   Age  ↔ Survived : -0.0772  (negatif sangat lemah)

💡 Kesimpulan:
   Semakin tinggi tarif tiket, sedikit lebih besar kemungkinan selamat.
   Fare sangat berkaitan dengan Pclass sebagai variabel utama.


### 📝 Interpretasi Q6

**6A — Kelompok Usia:**
- **Anak-anak (<12 th)** memiliki tingkat keselamatan tertinggi (~59%) → prioritas "children first".
- **Dewasa (18–59 th)** memiliki tingkat terendah (~38%) karena merupakan kelompok terbesar dengan persaingan sekoci paling ketat.

**6B — Ukuran Keluarga:**
- **Keluarga kecil (1–3 anggota)** → tingkat selamat tertinggi (~58%) → saling membantu saat evakuasi.
- **Pergi sendiri** → hanya ~30% selamat → tidak ada yang membantu koordinasi.
- **Keluarga sangat besar (7+)** → sangat sulit dievakuasi bersama, tingkat selamat sangat rendah.

**6C — Tarif Tiket:**
- Rata-rata tarif penumpang yang selamat lebih tinggi → berkorelasi positif lemah dengan keselamatan (Fare berkaitan erat dengan `Pclass`).

---
## 📋 Ringkasan Jawaban Semua Pertanyaan

| No. | Pertanyaan | Jawaban Singkat |
|-----|------------|-----------------|
| **Q1** | Dimensi & kolom dataset | **891 baris × 12 kolom** |
| **Q2** | Kolom dengan missing values | **Cabin** (77.10%), **Age** (19.87%), **Embarked** (0.22%) |
| **Q3** | % penumpang selamat | **38.38%** (342 dari 891) |
| **Q4** | Wanita kelas 1 | **94 orang**, rata-rata usia **34.61 th**, tingkat selamat **96.81%** |
| **Q5** | Kelas keselamatan tertinggi | **Kelas 1 (62.96%)** > Kelas 2 (47.28%) > Kelas 3 (24.24%) |
| **Q6** | Analisis tambahan | Usia, ukuran keluarga, dan tarif tiket memengaruhi peluang selamat |

---

### 🔑 Insight Utama

Tiga faktor paling berpengaruh terhadap keselamatan di Titanic:
1. **Jenis Kelamin** → Wanita jauh lebih banyak selamat (~74%) vs Pria (~19%)
2. **Kelas Tiket** → Kelas 1 (63%) vs Kelas 3 (24%) — 3× lipat perbedaannya
3. **Usia** → Anak-anak diprioritaskan dalam evakuasi

## Refleksi Singkat

Melalui praktikum pertemuan 2 ini saya berhasil menerapkan Pandas dan NumPy untuk menganalisis dataset nyata (Titanic). Saya memahami bagaimana `groupby()`, boolean indexing, dan fungsi agregasi bekerja untuk menjawab pertanyaan analitik. Temuan paling menarik bagi saya adalah betapa besar pengaruh kelas tiket dan jenis kelamin terhadap peluang selamat — yang merupakan cerminan kondisi sosial masyarakat awal abad ke-20.

---
*Notebook ini dibuat untuk keperluan akademik — Mata Kuliah Pengantar Data Science, Program Studi PJJ Informatika, Universitas UNSIA.*